repos/keg-cli/src/utils/services/publishService.js

```diff
@@ -0,0 +1,338 @@
+const { git } = require('KegGitCli')
+const { Logger } = require('KegLog')
+const { spawnCmd } = require('KegProc')
+const { ask } = require('@keg-hub/ask-it')
+const { copySync, emptyDirSync } = require('KegFileSys/fileSys')
+const { get } = require('@keg-hub/jsutils')
+const { getHubRepos } = require('../hub/getHubRepos')
+const { versionService } = require('./versionService')
+const { generalError } = require('../error/generalError')
+const { runRepoScript } = require('../hub/runRepoScript')
+const { getPublishContext } = require('../publish/getPublishContext')
+const { getPublishContextOrder } = require('../publish/getPublishContextOrder')
+const { getVersionUpdate, getValidSemver } = require('KegUtils/version')
+/**
+ * Attempts to rollback changes made to a repo using git
+ * @function
+ * @param {Object} repo - Repo to be rolled back
+ * @param {Object} publishArgs - Define the state or the repo being published
+ * @param {string} publishArgs.currentBranch
+ * @param {string} publishArgs.newVersion - i.e '1.0.0'
+ * @param {Boolean} publishArgs.wasPublished - whether yarn publish was executed
+ * @param {{number:Number, name:string}} publishArgs.step - current step information
+ * 
+ * @returns {Void}
+ */
+const rollbackChanges = async (repo, publishArgs) => {
+  const { currentBranch, newVersion, wasPublished, step } = publishArgs
+
+  logFormal(repo, `Publish service failed on step ${step.number}!\nRolling back publish changes...`)
+
+  const doGitReset = await ask.confirm(`Confirm running a full git reset. ALL CHANGES WILL BE LOST`)
+  if(!doGitReset) return Logger.warn(`Canceling git reset. Rollback did not complete. Current git branch is not clean!`)
+
+  if(step.number > 3 && wasPublished)
+    return Logger.warn(`\nCan not rollback changes, version ${newVersion} was already published to NPM!\n`)
+
+  if(!currentBranch)
+    return Logger.warn(`\nCan not rollback changes, Original git branch does not exist!\n`)
+
+  // Do a full git reset
+  logFormal(repo, `Resetting git to branch ${currentBranch}`)
+  await runGitCmd(`reset --hard HEAD`, repo.location)
```

Comment:
This works, but it also leaves us on the newly created publish branch
  * I.E. keg-8.0.1

We should reset the branch too. That way if we try to run the publish command again it doesn't already exist.
  * Otherwise the git branch updates will fail => http://snpy.in/B3GeNq

To do this:
  * Get the current branch of the repo and store it some where ( publishArgs? )
  * When `rollbackChanges` is called
    * Run the current reset steps as they are now
    * Switch back to the saved branch
    * Remove the publish branch if it was created

Reply:
this has been updated. 
alsol updated the testing instructions on rollback testing